File: sample-videochat-conference-java/src/main/java/com/quickblox/sample/videochat/conference/java/util/QBResRequestExecutor.java
package com.quickblox.sample.videochat.conference.java.util;

import android.os.Bundle;

import com.quickblox.chat.QBRestChatService;
import com.quickblox.chat.model.QBChatDialog;
import com.quickblox.chat.model.QBDialogType;
import com.quickblox.chat.request.QBDialogRequestBuilder;
import com.quickblox.chat.utils.DialogUtils;
import com.quickblox.core.QBEntityCallback;
import com.quickblox.core.exception.QBResponseException;
import com.quickblox.core.helper.StringifyArrayList;
import com.quickblox.core.request.QBPagedRequestBuilder;
import com.quickblox.core.request.QBRequestGetBuilder;
import com.quickblox.users.QBUsers;
import com.quickblox.users.model.QBUser;

import java.util.ArrayList;
import java.util.Collection;
import java.util.LinkedList;
import java.util.List;


public class QBResRequestExecutor {
    private String TAG = QBResRequestExecutor.class.getSimpleName();

    public void signUpNewUser(final QBUser newQbUser, final QBEntityCallback<QBUser> callback) {
        QBUsers.signUp(newQbUser).performAsync(callback);
    }

    public void signInUser(final QBUser currentQbUser, final QBEntityCallback<QBUser> callback) {
        QBUsers.signIn(currentQbUser).performAsync(callback);
    }

    public void signOut() {
        QBUsers.signOut().performAsync(null);
    }

    public void deleteCurrentUser(int currentQbUserID, QBEntityCallback<Void> callback) {
        QBUsers.deleteUser(currentQbUserID).performAsync(callback);
    }

    public void loadDialogs(final QBEntityCallback<ArrayList<QBChatDialog>> callback) {
        QBRequestGetBuilder requestBuilder = new QBRequestGetBuilder();
        requestBuilder.setLimit(100);
        QBRestChatService.getChatDialogs(QBDialogType.GROUP, requestBuilder).performAsync(callback);
    }

    public void loadDialogByID(String dialogId, final QBEntityCallback<QBChatDialog> callback) {
        QBRestChatService.getChatDialogById(dialogId).performAsync(callback);
    }


    public void deleteDialog(QBChatDialog qbDialog, QBEntityCallback<Void> callback) {
        QBRestChatService.deleteDialog(qbDialog.getDialogId(), false)
                .performAsync(callback);
    }

    public void deleteDialogs(Collection<QBChatDialog> dialogs, final QBEntityCallback<ArrayList<String>> callback) {
        StringifyArrayList<String> dialogsIds = new StringifyArrayList<>();
        for (QBChatDialog dialog : dialogs) {
            dialogsIds.add(dialog.getDialogId());
        }

        QBRestChatService.deleteDialogs(dialogsIds, false, null).performAsync(callback);
    }

    public void createDialogWithSelectedUsers(final List<QBUser> users, QBUser currentUser,
                                              final QBEntityCallback<QBChatDialog> callback) {

        QBRestChatService.createChatDialog(createDialog(users, currentUser)).performAsync(new QBEntityCallback<QBChatDialog>() {
            @Override
            public void onSuccess(QBChatDialog dialog, Bundle args) {
                callback.onSuccess(dialog, args);
            }

            @Override
            public void onError(QBResponseException responseException) {
                callback.onError(responseException);
            }
        });
    }

    private static QBChatDialog createDialog(List<QBUser> users, QBUser currentUser) {
        return DialogUtils.buildDialog(DialogUtils.createChatNameFromUserList(users.toArray(new QBUser[users.size()])),
                QBDialogType.GROUP, DialogUtils.getUserIds(users));
    }

    public void loadUsersByTag(final String tag, final QBEntityCallback<ArrayList<QBUser>> callback) {
        QBPagedRequestBuilder requestBuilder = new QBPagedRequestBuilder();
        requestBuilder.setPerPage(50);
        List<String> tags = new LinkedList<>();
        tags.add(tag);

        QBUsers.getUsersByTags(tags, requestBuilder).performAsync(callback);
    }

    public void loadUsersByIds(final Collection<Integer> usersIDs, final QBEntityCallback<ArrayList<QBUser>> callback) {
        QBUsers.getUsersByIDs(usersIDs, null).performAsync(callback);
    }

    public void loadUserById(int userID, QBEntityCallback<QBUser> callback){
        QBUsers.getUser(userID).performAsync(callback);
    }

    public void updateUser(QBUser qbUser, QBEntityCallback<QBUser> callback) {
        QBUsers.updateUser(qbUser).performAsync(callback);
    }

    public void updateDialog(QBChatDialog dialog, QBUser[] users, QBEntityCallback<QBChatDialog> callback) {
        QBDialogRequestBuilder requestBuilder = new QBDialogRequestBuilder();
        requestBuilder.addUsers(users);

        QBRestChatService.updateGroupChatDialog(dialog, requestBuilder).performAsync(callback);
    }
}

File: sample-videochat-conference-kotlin/src/main/java/com/quickblox/sample/videochat/conference/kotlin/util/QBResRequestExecutor.kt
package com.quickblox.sample.videochat.conference.kotlin.util

import android.os.Bundle
import com.quickblox.chat.QBRestChatService
import com.quickblox.chat.model.QBChatDialog
import com.quickblox.chat.model.QBDialogType
import com.quickblox.chat.request.QBDialogRequestBuilder
import com.quickblox.chat.utils.DialogUtils
import com.quickblox.core.QBEntityCallback
import com.quickblox.core.exception.QBResponseException
import com.quickblox.core.helper.StringifyArrayList
import com.quickblox.core.request.QBPagedRequestBuilder
import com.quickblox.core.request.QBRequestGetBuilder
import com.quickblox.sample.videochat.conference.kotlin.App
import com.quickblox.sample.videochat.conference.kotlin.R
import com.quickblox.users.QBUsers
import com.quickblox.users.model.QBUser
import java.util.*

private val connectionChecker: NetworkConnectionChecker = NetworkConnectionChecker(App.getInstance())

fun signUpNewUser(newQbUser: QBUser, callback: QBEntityCallback<QBUser>) {
    if (isInternetAvailable()) {
        QBUsers.signUp(newQbUser).performAsync(callback)
    } else {
        notifyInternetError(callback)
    }
}

fun signInUser(currentQbUser: QBUser, callback: QBEntityCallback<QBUser>) {
    if (isInternetAvailable()) {
        QBUsers.signIn(currentQbUser).performAsync(callback)
    } else {
        notifyInternetError(callback)
    }
}

fun signOut() {
    QBUsers.signOut().performAsync(null)
}

fun updateUser(qbUser: QBUser, callback: QBEntityCallback<QBUser>) {
    QBUsers.updateUser(qbUser).performAsync(callback)
}

fun deleteCurrentUser(currentQbUserID: Int, callback: QBEntityCallback<Void>) {
    if (isInternetAvailable()) {
        QBUsers.deleteUser(currentQbUserID).performAsync(callback)
    } else {
        notifyInternetError(callback)
    }
}

fun loadDialogs(callback: QBEntityCallback<ArrayList<QBChatDialog>>) {
    val requestBuilder = QBRequestGetBuilder()
    requestBuilder.limit = 100
    if (isInternetAvailable()) {
        QBRestChatService.getChatDialogs(QBDialogType.GROUP, requestBuilder).performAsync(callback)
    } else {
        notifyInternetError(callback)
    }
}

fun loadDialogByID(dialogId: String, callback: QBEntityCallback<QBChatDialog>) {
    if (isInternetAvailable()) {
        QBRestChatService.getChatDialogById(dialogId).performAsync(callback)
    } else {
        notifyInternetError(callback)
    }
}

fun deleteDialogs(dialogs: Collection<QBChatDialog>, callback: QBEntityCallback<ArrayList<String>>) {
    val dialogsIds = StringifyArrayList<String>()
    for (dialog in dialogs) {
        dialogsIds.add(dialog.dialogId)
    }
    if (isInternetAvailable()) {
        QBRestChatService.deleteDialogs(dialogsIds, false, null).performAsync(callback)
    } else {
        notifyInternetError(callback)
    }
}

fun createDialogWithSelectedUsers(users: List<QBUser>, currentUser: QBUser,
                                  callback: QBEntityCallback<QBChatDialog>) {

    QBRestChatService.createChatDialog(createDialog(users, currentUser)).performAsync(object : QBEntityCallback<QBChatDialog> {
        override fun onSuccess(dialog: QBChatDialog, args: Bundle) {
            callback.onSuccess(dialog, args)
        }

        override fun onError(responseException: QBResponseException) {
            callback.onError(responseException)
        }
    })
}

private fun isInternetAvailable(): Boolean {
    return connectionChecker.isConnectedNow()
}

private fun <T> notifyInternetError(callback: QBEntityCallback<T>?) {
    callback?.let {
        val error = App.getInstance().getString(R.string.no_internet_connection)
        val exception = QBResponseException(error)
        it.onError(exception)
    }
}

private fun createDialog(users: List<QBUser>, currentUser: QBUser): QBChatDialog {
    return DialogUtils.buildDialog(DialogUtils.createChatNameFromUserList(*users.toTypedArray()),
            QBDialogType.GROUP, DialogUtils.getUserIds(users))
}

fun loadUsersByTag(tag: String, callback: QBEntityCallback<ArrayList<QBUser>>) {
    val requestBuilder = QBPagedRequestBuilder()
    requestBuilder.perPage = 50
    val tags = LinkedList<String>()
    tags.add(tag)

    QBUsers.getUsersByTags(tags, requestBuilder).performAsync(callback)
}

fun loadUserById(userId: Int, callback: QBEntityCallback<QBUser>) {
    QBUsers.getUser(userId).performAsync(callback)
}

fun updateDialog(dialog: QBChatDialog, users: Array<QBUser>, callback: QBEntityCallback<QBChatDialog>) {
    val requestBuilder = QBDialogRequestBuilder()
    requestBuilder.addUsers(*users)

    if (isInternetAvailable()) {
        QBRestChatService.updateGroupChatDialog(dialog, requestBuilder).performAsync(callback)
    } else {
        notifyInternetError(callback)
    }
}